keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,472 @@
+However, the impact won't be significant due to the limitation on the number of pods per node.
+The evictions caused by this may be expensive depending on the number of pods and the requests conflicts and blocking PDBs.
+
+During the eviction, some pods may fail to be evicted. Users can use `pod_failed_eviction_node_affinity_total` metric to find them. 
+Typically, the reason for a failed eviction is due to a violation of the PDB. However, it is also possible that there could be issues with Kubernetes components. 
+In the case of the former, users need to manually check the relevant settings to ensure they are in compliance with the PDB. 
+If it is the latter, users should investigate the status of various Kubernetes components.
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeSelector struct {
+	NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeAffinity struct {
+	// If the affinity requirements specified by this field are not met at
+	// scheduling time, the pod will not be scheduled onto the node.
+	// If the affinity requirements specified by this field cease to be met
+	// at some point during pod execution (e.g. due to an update), the system
+	// will try to eventually evict the pod from its node.
+  // For v1alpha1, evictions will be handled by the Eviction API, 
+  // and will take the PDBs into consideration.
+	// +optional
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector
+}
+```
+
+Add a controller called `node-affinity-eviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node(excluding mirror pods), checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if `NodeSelector` still match the new node. 
+- If `RequiredDuringSchedulingRequiredDuringExecution` is no loger met, trying to evict the pod.
+
+- For alpha stage, we'll use a simple eviction API to evict Pods.
+  - It will use Eviction API to evict the pod(The API takes the PDBs into consideration, so we do not have to use the PDBs explicitly here). 
+  - If eviction fails, we will take different actions based on the failure reasons:
+    - If eviction fails due to a violation of the PDB, we will keep the target Pod unchanged. 
+    - If eviction fails for any other reason, we will attempt a certain number of retries. If the eviction continues to fail after multiple retries, we will keep the target Pod unchanged. 
+  - If a pod fails to be evicted during the current eviction attempt, we will retry the eviction the next time the labels of that node change.
+
+- After the eviction finally fails, we'll print a log to recored the event. And increase the metric `pod_failed_eviction_node_affinity_total` labeled with pod's name. So that users can find out this and take care of the pod manually.
```

Comment:
how do we recognize a final failure?

Reply:
The eviction API returned a 429 status, or it continued to return a non-200 status even after multiple retries?

Reply:
what do we gain from distinguishing the errors?